pallets/subtensor/src/subnets/weights.rs

```diff
@@ -18,34 +18,45 @@ impl<T: Config> Pallet<T> {
     ///   - The hash representing the committed weights.
     ///
     /// # Raises:
-    /// * `WeightsCommitNotAllowed`:
-    ///   - Attempting to commit when it is not allowed.
+    /// * `CommitRevealDisabled`:
+    ///   - Attempting to commit when the commit-reveal mechanism is disabled.
     ///
+    /// * `TooManyUnrevealedCommits`:
+    ///   - Attempting to commit when the user has more than the allowed limit of unrevealed commits.
     pub fn do_commit_weights(
         origin: T::RuntimeOrigin,
         netuid: u16,
         commit_hash: H256,
     ) -> DispatchResult {
+        // --- 1. Check the caller's signature (hotkey).
         let who = ensure_signed(origin)?;
 
         log::debug!("do_commit_weights( hotkey:{:?} netuid:{:?})", who, netuid);
 
+        // --- 2. Ensure commit-reveal is enabled for the network.
         ensure!(
             Self::get_commit_reveal_weights_enabled(netuid),
             Error::<T>::CommitRevealDisabled
         );
 
-        ensure!(
-            Self::can_commit(netuid, &who),
-            Error::<T>::WeightsCommitNotAllowed
-        );
+        // --- 3. Check if the current number of unrevealed commits is within the allowed limit.
+        WeightCommits::<T>::try_mutate(netuid, &who, |maybe_commits| -> DispatchResult {
+            if let Some(ref commits) = maybe_commits {
+                ensure!(commits.len() < 10, Error::<T>::TooManyUnrevealedCommits);
```

Comment:
Attempting to commit should remove expired commits from the queue, otherwise if all of them expire, you might brick a hotkey (and also lets not write the expired commits back to the storage)